armi/reactor/blocks.py

```diff
@@ -1580,61 +1580,80 @@ def setPinPowers(
+            The block-level pin linear power densities. powers[i] represents the average
+            linear power density of pin i. The units of linear power density is watts/cm
+            (i.e., watts produced per cm of pin length). The "ARMI pin ordering" must be
+            be used, which is counter-clockwise from 3 o'clock.
 
-        Notes
-        -----
-        This handles rotations using the pinLocation parameters.
+        numPins : int, required
+            Number of pins in the block. This parameter should probably be removed because it
+            can be derived from the block.
 
-        This sets:
+        imax: int, required
+            Number of pin rings. This parameter should probably be removed because it
+            can be derived from the block.
 
-        self.p.pinPowers : 1-D numpy array
-            The block-level pin linear power densities. pinPowers[i] represents the average linear
-            power density of pin i.
-            Power units are Watts/cm (Watts produced per cm of pin length).
-            The "ARMI pin ordering" is used, which is counter-clockwise from 3 o'clock.
+        jmax: list of ints, required
+            Number of pins per ring. Element 0 is the number of pins in ring 1, element 1
+            is the number of pins in ring 2, and so on. This parameter should probably be
+            removed because it can be derived from the block.
+
+        removeSixCornerPins: bool, optional.
+            Defaults to False
+
+        powerKeySuffix: str, optional
+            Must be either an empty string, :py:const:`NEUTRON <armi.physics.neutronics.const.NEUTRON>`,
+            or :py:const:`GAMMA <armi.physics.neutronics.const.GAMMA>`. Defaults to empty
+            string.
+
+        Notes
+        -----
+        This method can handle assembly rotations by using the *pinLocation* parameter.
         """
-        self.p.pinPowers = numpy.zeros(numPins)
-        self.p["linPowByPin" + powerKeySuffix] = numpy.zeros(numPins)
+        powerKey = f"linPowByPin{powerKeySuffix}"
+        self.p[powerKey] = numpy.zeros(numPins)
+
+        # Loop through rings
         j0 = jmax[imax - 1] / 6
         pinNum = 0
-        for i in range(imax):  # loop through rings
-            for j in range(jmax[i]):  # loop through positions in ring i
+        for i in range(imax):
+            # Loop through positions in ring i
+            for j in range(jmax[i]):
                 if removeSixCornerPins and i == imax - 1 and math.fmod(j, j0) == 0.0:
                     linPow = 0.0
                 else:
+                    # TODO: This appears to need fixing to account for blocks in fueled
+                    # TODO: assemblies that contain elevations with pins but no fuel,
+                    # TODO: such as for an axial shield. These blocks should still have
+                    # TODO: powers that are eligible for rotation.
                     if self.hasFlags(Flags.FUEL):
```

Comment:
Generally, if a PR comes with a `TODO`, that means it needs a little more work until it's ready for prime-time.

"fueled assembles that contain elevations with pins but no fuel"

@ntouran  Do you understand why this if-statement exists?

Reply:
I added it so that it could be discussed as part of the PR. The if/else doesn't make sense to me.

Reply:
In doing a line blame, it looks like this if/else has been here for at least 3 years...I don't know how to go beyond that time period because it looks like the file was refactored from somewhere else, so that commit shows one gigantic add operation to the file.

Again, my confusion here is that the line says "if you are fuel, then do a rotation lookup before getting your pin number; otherwise, don't do a lookup." I don't understand why you would want to do that ;why does it matter whether you are fuel or not?

Reply:
I'm also confused here; I can't think of a reason why only the fuel blocks should be rotated. I think this if/else is unnecessary and/or wrong.

Reply:
I think that you're right that this is probably wrong. The `blocks` module has `PIN_COMPONENTS` defined as a module-level variable. You might be able to see any children of the block have flags that match these and if so then the block contains "pins". 

Honestly, I think that the concept of "pins" needs to be standardized because I can see that there are places in the code that assume if "mult" > 1 then it's a pin and then there's other places that now use this `PIN_COMPONENTS` list. This seems sketchy and fragile.

@ntouran - See question above by @john-science please. I don't understand this either.